### ISSUE TITLE
AOT support

### ISSUE THREAD

**Issue by mpicard** [maximusk/extension-mechanism-demo#5]
Hi is there a way to get the "platform" building with AOT? I tried myself but I am getting this error message: `"Unexpected value 'Lt' imported by the module 'WidgetModule'. Please add a @NgModule annotation."`

I had to add some providers to the app as follows:

```
export function createCompiler(compilerFactory: CompilerFactory) {
  return compilerFactory.createCompiler();
}

...

  providers: [
    { provide: COMPILER_OPTIONS, useValue: {}, multi: true },
    {
      provide: CompilerFactory,
      useClass: JitCompilerFactory,
      deps: [COMPILER_OPTIONS]
    },
    { provide: Compiler, useFactory: createCompiler, deps: [CompilerFactory] }
  ]
```

**Comment by mpicard** [comment]
I managed to figure it out! Here is a POC https://github.com/mpicard/angular-dynamic-remote-modules that I created taking inspiration from yourself @maximusk  and my colleague @paucls

**Comment by mkoretskyi** [comment]
@mpicard , great job!

**Comment by maximusk** [comment]
@mpicard , great job!

**Comment by BuckyMaler** [comment]
@maximusk and @mpicard thank you both for sharing your work concerning loading modules at runtime. Max, your [talk](https://www.youtube.com/watch?v=pERhnBBae2k&t=0s&index=3&list=PLw5h0DiJ-9PDyHiRwuP2c_kuxx1URzUMx) provided the initial fuel to dive deeper into this topic. Martin, I benefited by poking at your  [modular-angular-framework](https://gitlab.com/modular-angular-framework).

At work, I've been tasked with facilitating individual deployments for numerous "sub-apps" that make up one large "portal" app. At this time, all sub-apps are bundled as one large app, but for several reasons we want each app to be deployable separately. With inspiration from your work, I have a PoC using SystemJS to load the sub-apps, which are UMD files bundled by webpack, based on route. This is also using Angular's JIT compiler.

We currently bundle with AOT and would prefer to continue to benefit from AOT. Martin, I get the error you've noted above after running `ng build --prod` and serving all apps in the `modular-angular-framework` repo. Were you able to AOT compile the "shell" app?

Are either of you aware of discussion or work concerning AOT compiling sub-apps, or extensions, as Max puts it?

Again, thank you both for sharing your knowledge. Max, I hope you don't mind my posting of not an actual issue and continuing discussion in a closed issue.

**Comment by maximusk** [comment]
@BuckyMaler hey, I'm glad my talk inspired you. With my application, I never made to the stage where I'd build shell apps with AOT and them load them. Do you get the error at runtime or build time?

**Comment by BuckyMaler** [comment]
@maximusk with all distributions bundled with AOT enabled, I get a runtime error. The shell app loads as expected, but the sub-apps, which require a bit of nuance when bundling, throw a runtime error. From my understanding, Angular is trying to JIT compile the lazily loaded sub-app, but since the shell app was AOT compiled the JIT compiler isn't shipped to the browser. Since Angular is reaching for the JIT compiler, I assume the sub-apps aren't properly AOT compiled.

<img width="628" alt="screen shot 2018-11-17 at 4 19 50 pm" src="https://user-images.githubusercontent.com/14145352/48667978-5c5ffd00-eaa8-11e8-8e6d-5782e433c980.png">

I created an [example repo](https://github.com/BuckyMaler/subapp-mechanism-example) if you have time to take a look. Code related to lazily loading the sub-app can be found in `apps/shell/src/app/app.module.ts`, and config related to bundling the sub-app can be found in `angular.json` and `tools/hooks/subapp-mod.ts`. The hook is part of a custom Angular CLI builder that ensures a single UMD bundle is output for the sub-app.

The example repo is in a working state. To **reproduce** the runtime error, pop in `angular.json` and change `"aot": false` to `"aot": true` for both projects.

I noticed that changing webpack's main entry point in `angular.json` for the sub-app from `main.ts` to `app.module.ts` resulted in an empty `AppModule` in the outputted bundle with AOT enabled. Furthermore, no decorator metadata or `.ngfactory` are created. However, I changed the entry point because I cannot bootstrap the sub-apps as bootstrapping is a responsibility of the shell app.
